pkg/ipc/ipc.go

```diff
@@ -359,10 +362,18 @@ func (env *Env) parseOutput(p *prog.Prog) (*ProgInfo, error) {
 			return nil, fmt.Errorf("call %v/%v/%v: signal overflow: %v/%v",
 				i, reply.index, reply.num, reply.signalSize, len(out))
 		}
-		if inf.Cover, ok = readUint32Array(&out, reply.coverSize); !ok {
+		var rawCover []uint32
+		if rawCover, ok = readUint32Array(&out, reply.coverSize); !ok {
 			return nil, fmt.Errorf("call %v/%v/%v: cover overflow: %v/%v",
 				i, reply.index, reply.num, reply.coverSize, len(out))
 		}
+		rawCover = append([]uint32{}, rawCover...)
```

Comment:
Why is this copy necessary?
Can't this be just:
```
		if inf.RawCover, ok = readUint32Array(&out, reply.coverSize); !ok {
			...
		}
		inf.Cover = inf.RawCover
		if env.config.DedupCover {
			inf.Cover = dedupCover(inf.Cover)
		}
```
?

Reply:
From the code, it doesn't really seem reliable. `readUint32Array` just points the slice to a subpart of the shared buffer (and it is preserved between the executions within the same `Env`).

And we have to deal with it in other parts of code. `Cover` is copied immediately after execution here https://github.com/google/syzkaller/blob/master/syz-fuzzer/proc.go#L142

And signal here: https://github.com/google/syzkaller/blob/master/syz-fuzzer/proc.go#L265

It looks fragile even now. I fear that with 3 such slices it will get even more prone to errors.

Reply:
But it worked before. Don't add such subtle changes with separate justifications here.